src/main/java/org/kuali/kfs/module/cam/businessobject/AssetLocationGlobal.java

```diff
@@ -0,0 +1,252 @@
+     * address ( building, and room) and an off campus addresses. If the building and room are updated from location
+     * global then the following fields should be set to null
+     * WHERE the ast_loc_typ_cd = 'O' cm_ast_loc_t.ast_loc_cntnt_nm cm_ast_loc_t.ast_loc_strt_addr
+     * cm_ast_loc_t.ast_loc_city_nm cm_ast_loc_t.ast_loc_state_cd cm_ast_loc_t.ast_loc_cntry_cd
+     * cm_ast_loc_t.ast_loc_zip_cd
+     *
+     * @param asset
+     */
+    private void updateOffCampusWithOnCampusValues(final Asset asset) {
+        if (asset.getAssetLocations() != null) {
+            for (final AssetLocation location : asset.getAssetLocations()) {
+                final boolean offCampus = CamsConstants.AssetLocationTypeCode.OFF_CAMPUS.equals(
+                        location.getAssetLocationTypeCode());
+                final boolean buildingOrRoom = StringUtils.isNotBlank(asset.getBuildingCode())
+                                               || StringUtils.isNotBlank(asset.getBuildingRoomNumber());
+                if (offCampus && buildingOrRoom) {
+                    location.setAssetLocationContactName(null);
+                    location.setAssetLocationStreetAddress(null);
+                    location.setAssetLocationCityName(null);
+                    location.setAssetLocationStateCode(null);
+                    location.setAssetLocationCountryCode(null);
+                    location.setAssetLocationZipCode(null);
+                }
+            }
+        }
+    }
+
+    /*
+     * CU Customization: Modified isPersistable() implementation to prevent the document from being saved
+     * when any of the detail objects specify an invalid building or room. Also added related helper methods.
+     * 
+     * Some of the code and logic here has been duplicated from the AssetLocationGlobalRule class.
+     */
+
+    @Override
+    public boolean isPersistable() {
+        boolean canPersist = true;
+        if (CollectionUtils.isNotEmpty(assetLocationGlobalDetails)) {
+            int index = 0;
+            for (final AssetLocationGlobalDetail detail : assetLocationGlobalDetails) {
+                final String errorPath = StringUtils.join(KRADConstants.MAINTENANCE_NEW_MAINTAINABLE,
+                        CamsPropertyConstants.AssetLocationGlobal.ASSET_LOCATION_GLOBAL_DETAILS, "[", index, "]");
+                GlobalVariables.getMessageMap().addToErrorPath(errorPath);
+                canPersist &= detailHasValidOrIncompleteForeignKeyValues(detail);
+                GlobalVariables.getMessageMap().removeFromErrorPath(errorPath);
+                index++;
+            }
+        }
+        return canPersist;
+    }
+
+    private boolean detailHasValidOrIncompleteForeignKeyValues(final AssetLocationGlobalDetail detail) {
+        boolean valid = true;
+        final boolean hasCampusCode = StringUtils.isNotBlank(detail.getCampusCode());
```

Comment:
do we need to check  if there is just a campus code and the campus code is wrong?
What if there is a building code, but no campus code?
what is there is a room number but no building code?

Reply:
I was only encountering the unexpected exception when one of the database foreign key constraints was not being satisfied. The two problematic constraints in this case were for the Building (using Campus Code and Building Code) and the Room (using Campus Code, Building Code and Room Number); this document's BOs don't appear to have a foreign key constraint for just the Campus. Also, as long as at least one of the fields involved in the foreign key constraint is null/blank, the constraint won't be validated by the database when saving the document.